float8_experimental/float8_dynamic_linear.py

```diff
@@ -31,13 +32,23 @@ def forward(
 
     @staticmethod
     def backward(ctx, gradY):
-        gradY_scale = tensor_to_scale(gradY, torch.float8_e5m2)
-        gradY_scaled = gradY * gradY_scale
-        bits_fp8 = to_fp8_saturated(gradY_scaled, torch.float8_e5m2)
-        return (
-            Float8Tensor(bits_fp8, gradY_scale, gradY.dtype, emulate=ctx.emulate),
-            None,
-        )
+        fp8_tensor = to_fp8_no_autograd(gradY, torch.float8_e5m2, ctx.emulate)
+        return fp8_tensor, None
+
+
+def cast_x_to_float8_e4m3fn_pre_hook(module, args):
+    """
+    Hook to cast the incoming activation to `torch.float8_e4m3fn`
+    """
+    return module.cast_to_float8_e4m3fn(args[0])
+
+
+def cast_dldy_to_float8_e5m2_backward_pre_hook(module, grad_output):
```

Comment:
testing my understanding: 

It looks to me that the torch.compile issue is related to the backward_prehook and tensor subclass interactions. 

But could this be solved by using `module.register_forward_hook` and then inside the forward hook we call `y = self.cast_to_float8_e5m2_bw(y)` just like the current casting?

This would solve the backward prehook issue I believe, but not sure if it would hit subclass issues (hopefully not)

Reply:
ahhh great, idea unfortunately still erroring for compile..

I think that is a little cleaner, is that enough for a good interaction with DTensor?

Reply:
Yeah I think a forward_hook might be a little cleaner and easier for DTensor to interact. We would need to actually try composing those hooks to see if there's any gap. I feel let's try land the forward_hook approach in this PR, and if we found we need `full_backward_pre_hook` instead of `forward_hook` later, we can always change it back later if needed?

Reply:
sounds good!